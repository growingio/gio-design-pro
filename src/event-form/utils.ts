/* eslint-disable no-prototype-builtins */
/* eslint-disable consistent-return */

/**
 * 表单数据最大长度
 */
export const MAX_VALUE_LENGTH = 30;
/**
 * 描述字段最大长度
 */
export const MAX_DESC_LENGTH = 150;

export interface Kv {
  key: string;
  value: string;
  [K: string]: any;
}
/**
 * 去除字符串字段中前后的空格
 * @param val
 */
export function trim(val: any): any {
  if (!val) {
    return val;
  }

  if (val.trim) {
    return val.trim.call(val);
  }

  if (Array.isArray(val)) {
    return val.map((v) => trim(v));
  }

  if (typeof val === 'object') {
    const obj = {} as any;
    Object.keys(val).forEach((key) => {
      obj[trim(key)] = trim(val[key]);
    });
    return obj;
  }

  return val;
}

export const isEmptyStr = (val: string) => {
  if (!val) {
    return true;
  }
  return val?.trim().length === 0;
};

export const isObject = (o: any) => Object.prototype.toString.call(o) === '[object Object]';

export const isEmptyValue = (val: any) => val === null || val === undefined || val === '';

export const kvsToQuery = (kvs: Kv[]): any => {
  if (!kvs || kvs.length === 0) return;
  return kvs
    .map((kv) => {
      if (kv?.key) {
        return `${kv?.key}=${kv?.value || ''}`;
      }
      return '';
    })
    .filter((v) => v)
    .join('&');
};

export const objToQuery = (obj: any) =>
  Object.keys(obj || {})
    .map((key) => `${key}=${obj[key]}`)
    .filter((v) => v)
    .join('&');

// 查询参数转化到kv结构数组，过滤掉key不存在的
export const queryToKvs = (query?: string): Kv[] => {
  if (!query) return [];
  // eslint-disable-next-line no-param-reassign
  if (query.startsWith('?')) query = query.slice(1);
  return query
    .split('&')
    .map((kv) => {
      const kva = kv.split('=');
      return {
        key: kva[0],
        value: kva[1],
      };
    })
    .filter((v) => !!v.key);
};
export type Query = { [name: string]: string | any[] };
export const qsParse = (search: string): Query => {
  let _search = search;
  if (search?.startsWith('?')) _search = search.slice(1);
  const query = {} as any;
  _search?.split('&').forEach((item) => {
    const kv = item.split('=');
    if (query.hasOwnProperty(kv[0])) {
      let v = query[kv[0]];
      if (Array.isArray(v)) {
        v.push(kv[1]);
      } else {
        v = [v, kv[1]];
      }
      query[kv[0]] = v?.sort();
    } else {
      !isEmptyStr(kv[0]) && (query[kv[0]] = kv[1] || '');
    }
  });
  return query;
};
/**
 * 判断某个元素是不是一个元素的子元素
 * @param elem 目标元素
 * @param parentElem 父元素
 */
export const isChildOfElement = (elem: Element, parentElem: Element) => {
  let p = elem.parentElement;
  while (p && p !== document.body) {
    if (p === parentElem) {
      return true;
    }
    p = p.parentElement;
  }
  return false;
};
export const matchString = (likePattern: string, value: string, accurate = false): boolean => {
  if (!accurate && (!likePattern || value == null)) {
    return false;
  }
  return !accurate && likePattern.indexOf('*') !== -1
    ? new RegExp(`^${likePattern.replace(/\*+/g, '(.*)')}$`).test(value)
    : likePattern === value;
};
/**
 * 模糊模式下只要page中包含tag中的所有query就行，多了是可以匹配的
 * a=1&b=1可以匹配上a=1
 * 不关注顺序
 *
 * @param patternSearch 匹配模式
 * @param pageSearch 要匹配的query
 * @param accurate 开启精准匹配，参数key一致，值相同，默认false
 */
export const matchQuery = (patternSearch: string, pageSearch: string, accurate = false): boolean => {
  const patternQuery = qsParse(patternSearch);
  const pageQuery = qsParse(pageSearch);

  // 精准匹配下，query个数必须一样
  if (accurate && Object.keys(patternQuery).length !== Object.keys(pageQuery).length) {
    return false;
  }
  // page中的query在多的情况下也可匹配成功
  return Object.keys(patternQuery).every((key) => {
    // eslint-disable-next-line no-prototype-builtins
    if (!pageQuery.hasOwnProperty(key)) {
      return false;
    }
    const pageVal = pageQuery[key] as any;
    const patternVal = patternQuery[key];
    // 匹配规则和页面参数值类型不一样就直接匹配失败,数组和字符串
    if (typeof patternVal !== typeof pageVal) {
      return false;
    }
    if (Array.isArray(patternVal)) {
      // 在精准匹配模式下，数组内容必须是一样的
      if (Array.isArray(pageVal) && (accurate ? pageVal.length === patternVal.length : true)) {
        // 对于数组的参数，比较时不分顺序进行匹配
        return patternVal.every((ptv) => Array.from(pageVal).some((pgv: any) => matchString(ptv, pgv, accurate)));
      }
      return false;
    }
    return matchString(patternVal, pageVal, accurate);
  });
};
