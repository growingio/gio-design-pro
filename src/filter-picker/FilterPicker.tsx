import React, { useState } from 'react';
import { Dropdown, Button } from '@gio-design/components';
import { FilterOutlined } from '@gio-design/icons';
import FilterOverlay from './components/FilterOverlay/index';
import './style/index.less';

import { FilterPickerProps, FilterValueType, operationsOptionType } from './interfaces';

export const FilterPickerContext = React.createContext<Pick<FilterPickerProps, 'fetchDetailData' | 'operationsOption'>>(
  {}
);

const defaultOperationsOption: operationsOptionType = {
  string: ['=', '!=', 'in', 'not in', 'like', 'not like', 'hasValue', 'noValue'],
  int: ['=', '!=', '>', '>=', '<', '<=', 'between', 'not between', 'hasValue', 'noValue'],
  date: ['=', '!=', '>', '<', 'relativeBetween', 'relativeCurrent', 'between', 'not between', 'hasValue', 'noValue'],
  STRING: ['=', '!=', 'in', 'not in', 'like', 'not like'],
};

const FilterPicker = (props: FilterPickerProps) => {
  const {
    children,
    getTooltipContainer,
    filter,
    propertyOptions,
    onConfirm,
    dimensionValueRequest,
    measurements,
    timeRange,
    recentlyStorePrefix,
    fetchDetailData,
    operationsOption,
    hasVisible = false,
    visible,
    onVisibleChange,
    placement = 'bottomRight',
  } = props;
  const [localVisible, setLocalVisible] = useState(false);
  const visibleChange = (v: boolean) => {
    setLocalVisible(v);
    onVisibleChange?.(v);
  };
  const cancel = () => {
    setLocalVisible(false);
    hasVisible && onVisibleChange?.(false);
  };
  const submit = (v: FilterValueType[]) => {
    setLocalVisible(false);
    hasVisible && onVisibleChange?.(false);
    onConfirm({ ...filter, exprs: v });
  };
  return (
    <FilterPickerContext.Provider
      value={{ fetchDetailData, operationsOption: { ...defaultOperationsOption, ...operationsOption } }}
    >
      <Dropdown
        visible={hasVisible ? visible : localVisible}
        trigger={['click']}
        onVisibleChange={visibleChange}
        overlay={
          <FilterOverlay
            onCancel={cancel}
            onSubmit={submit}
            filterList={filter.exprs}
            propertyOptions={propertyOptions}
            dimensionValueRequest={dimensionValueRequest}
            measurements={measurements}
            timeRange={timeRange}
            recentlyStorePrefix={recentlyStorePrefix}
          />
        }
        placement={placement}
        getTooltipContainer={getTooltipContainer}
        destroyTooltipOnHide
      >
        {children || <Button icon={<FilterOutlined />} size="small" type={!localVisible ? 'link' : 'secondary'} />}
      </Dropdown>
    </FilterPickerContext.Provider>
  );
};

export default FilterPicker;
