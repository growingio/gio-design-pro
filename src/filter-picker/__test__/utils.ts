/* eslint-disable import/prefer-default-export */
import { act } from 'react-dom/test-utils';

const globalTimeout = global.setTimeout;

export const sleep = async (timeout = 0) => {
  await act(async () => {
    await new Promise((resolve) => globalTimeout(resolve, timeout));
  });
};
