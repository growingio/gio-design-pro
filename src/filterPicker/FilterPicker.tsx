import React, { useState } from 'react';
import { Dropdown, Button } from '@gio-design-new/components';
import { FilterOutlined } from '@gio-design/icons';
import FilterOverlay from './components/FilterOverlay/index';
import './style/index.less';

import { FilterPickerProps, FilterValueType } from './interfaces';

const FilterPicker = (props: FilterPickerProps) => {
  const {
    children,
    getTooltipContainer,
    filter,
    propertyOptions,
    onConfirm,
    dimensionValueRequest,
    measurements,
    timeRange,
  } = props;
  const [visible, setVisible] = useState(false);

  const visibleChange = (v: boolean) => {
    setVisible(v);
  };
  const cancel = () => {
    setVisible(false);
  };
  const submit = (v: FilterValueType[]) => {
    setVisible(false);
    onConfirm({ ...filter, exprs: v });
  };
  return (
    <Dropdown
      visible={visible}
      trigger={['click']}
      onVisibleChange={visibleChange}
      overlay={
        <FilterOverlay
          onCancel={cancel}
          onSubmit={submit}
          filterList={filter.exprs}
          propertyOptions={propertyOptions}
          dimensionValueRequest={dimensionValueRequest}
          measurements={measurements}
          timeRange={timeRange}
        />
      }
      placement="bottomRight"
      getTooltipContainer={getTooltipContainer}
      destroyTooltipOnHide
    >
      {children || (
        <Button icon={<FilterOutlined />} size="small" type="assist">
          过滤条件
        </Button>
      )}
    </Dropdown>
  );
};

export default FilterPicker;
