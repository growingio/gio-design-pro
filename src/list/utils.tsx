/* eslint-disable import/prefer-default-export */
import React from 'react';
import usePrefixCls from '@gio-design/components/es/utils/hooks/use-prefix-cls';
import Item from './Item';
import ExpandItem from './ExpandItem';
import { ListItemProps } from './interfaces';

const DEFAULT_SHOW_ITEMS_COUNT = 10;

export const rootPrefixCls = () => usePrefixCls('list');

export function renderItem(i: ListItemProps, idx: number) {
  return <Item {...i} key={`item-${i.key}-${idx}`} />;
}

export function renderExpandableItems(expanded: boolean, currentItems: ListItemProps[], onExpand: () => void) {
  if (expanded) {
    return currentItems.map(renderItem);
  }
  if (currentItems.length > DEFAULT_SHOW_ITEMS_COUNT) {
    const showItems = currentItems.slice(0, DEFAULT_SHOW_ITEMS_COUNT);
    return showItems
      .map(renderItem)
      .concat(
        <ExpandItem
          title={`展开全部 (${currentItems.length - showItems.length})`}
          key={`expand-item-${showItems.length + 1}`}
          onClick={onExpand}
        />
      );
  }
  return currentItems.map(renderItem);
}
