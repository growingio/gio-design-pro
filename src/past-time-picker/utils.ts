import has from 'lodash/has';
import get from 'lodash/get';
import { format, startOfToday, sub } from 'date-fns';
import { TimeMode } from './interfaces';
import { DATE_FORMAT } from './constant';

export const parseTimeMode = (timeRange: string | undefined, QUICK_MAPPING: any) => {
  if (!timeRange) {
    return undefined;
  }
  if (has(QUICK_MAPPING, timeRange)) {
    return 'quick';
  }
  const items = timeRange.split(':');
  switch (items[0]) {
    case 'since':
      return TimeMode.Since;
    case 'abs':
      return TimeMode.Absolute;
    case 'day':
      return TimeMode.Relative;
    default:
      return undefined;
  }
};

export const parseStartAndEndDate = (timeRange: string | undefined): [Date | undefined, Date | undefined] => {
  if (!timeRange || timeRange.split(':').length !== 2) {
    return [undefined, undefined];
  }
  const items = timeRange.split(':');
  const times = items[1].split(',').map((str) => parseInt(str, 10));
  const today = startOfToday();
  if (items[0] === 'since') {
    if (times.length === 1) {
      return [new Date(times[0]), today];
    }
    return [new Date(times[0]), sub(today, { days: times[1] })];
  }
  if (items[0] === 'abs') {
    return [new Date(times[0]), new Date(times[1])];
  }
  if (items[0] === 'day') {
    return [sub(today, { days: times[0] }), sub(today, { days: times[1] })];
  }
  return [undefined, undefined];
};

export const parseFixedMode = (timeRange: string | undefined) => {
  if (!timeRange || timeRange.split(':').length !== 2) {
    return false;
  }
  const items = timeRange.split(':');
  const times = items[1].split(',');
  if (items[0] === 'since') {
    if (times.length === 1 || times[1] === '0') {
      return 'today';
    }
    return 'yesterday';
  }
  if (items[0] === 'day' && times[1] === '1') {
    return 'yesterday';
  }
  return false;
};

export const humanizeTimeRange = (
  timeRange: string,
  defaultString: string = '时间范围',
  QUICK_MAPPING: any,
  locale: any
) => {
  const { FromText, toTodayText, toYesterdayText, lastText, dayText, ToText } = locale;
  if (!timeRange || timeRange.split(':').length !== 2) {
    return defaultString;
  }
  if (has(QUICK_MAPPING, timeRange)) {
    return get(QUICK_MAPPING, timeRange);
  }
  const items = timeRange.split(':');
  const times = items[1].split(',').map((str) => parseInt(str, 10));
  if (items[0] === 'since') {
    const start = format(times[0], DATE_FORMAT);
    if (times.length === 1) {
      return `${FromText} ${start} ${toTodayText}`;
    }
    return `${FromText} ${start} ${toYesterdayText}`;
  }
  if (items[0] === 'abs') {
    const start = format(times[0], DATE_FORMAT);
    const end = format(times[1], DATE_FORMAT);
    return `${FromText} ${start} ${ToText} ${end}`;
  }
  if (items[0] === 'day') {
    if (times[1] === 1) {
      return `${lastText} ${times[0] - times[1]} ${dayText}`;
    }
    return `${lastText} ${times[1]}-${times[0]} ${dayText}`;
  }
  return defaultString;
};

export default {
  parseTimeMode,
  parseStartAndEndDate,
  humanizeTimeRange,
};
