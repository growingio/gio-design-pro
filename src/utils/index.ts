import React from 'react';
import { Resource } from './interfaces';

export function resourceToItem(
  resource: Resource,
  mouseEvents?: {
    onClick?: (e: React.MouseEvent<HTMLElement>) => void;
    onMouseEnter?: (e: React.MouseEvent<HTMLElement>) => void;
    onMouseLeave?: (e: React.MouseEvent<HTMLElement>) => void;
  },
  disabled?: boolean
) {
  return {
    key: resource.id,
    children: resource.name,
    ...mouseEvents,
    disabled,
  };
}

export default {
  resourceToItem,
};
