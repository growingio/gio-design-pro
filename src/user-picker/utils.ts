import { Option } from '@gio-design/components/es/components/list/interface';
import { Segment } from '../types';
import { OptionGroup, PreparedSegment } from './interfaces';

export function segmentToOption(seg: PreparedSegment | Segment, group?: OptionGroup): Option {
  if (!group) {
    return {
      value: seg.id,
      label: seg.name,
    };
  }
  return {
    value: seg.id,
    label: seg.name,
    groupLabel: group.label,
    groupValue: group.key,
  };
}

export default {
  segmentToOption,
};
