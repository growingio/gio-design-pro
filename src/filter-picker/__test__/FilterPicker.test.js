import React from 'react';
import { mount, render } from 'enzyme';
import FilterPicker from '..';
import { sleep } from './utils.ts';

const originalOffsetHeight = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetHeight');
const originalOffsetWidth = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetWidth');
beforeEach(() => {
  // For <List /> use <AutoSizer />.
  // <AutoSizer /> uses offsetWidth and offsetHeight.
  // Jest runs in JSDom which doesn't support measurements APIs.
  // @see https://github.com/testing-library/react-testing-library/issues/353#issuecomment-510046921
  Object.defineProperty(HTMLElement.prototype, 'offsetHeight', { configurable: true, value: 320 });
  Object.defineProperty(HTMLElement.prototype, 'offsetWidth', { configurable: true, value: 160 });

  jest.useFakeTimers();
});

afterEach(() => {
  Object.defineProperty(HTMLElement.prototype, 'offsetHeight', originalOffsetHeight);
  Object.defineProperty(HTMLElement.prototype, 'offsetWidth', originalOffsetWidth);

  jest.useRealTimers();
});

describe('FilterPicker', () => {
  // 基础渲染
  const getFilterPicker = (props) => (
    <FilterPicker
      propertyOptions={[
        {
          id: 'test_int_id',
          name: 'test_int_name',
          groupId: 'user',
          groupName: '用户属性',
          type: 'usr',
          valueType: 'int',
          __typename: 'Dimension',
        },
      ]}
      filter={{ op: 'and', exprs: [], __typename: 'directivesFilter' }}
      {...props}
    />
  );

  it('should match FilterPicker base snapshot.', () => {
    const wrapper = render(getFilterPicker());
    expect(wrapper).toMatchSnapshot();
  });

  // 使用 children 时，应该正确显示
  it('when use children, should render it correctly', () => {
    const children = <div>过滤选择器</div>;
    const wrapper = render(getFilterPicker({ children }));
    expect(wrapper.text()).toEqual('过滤选择器');
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const wrapper = mount(getFilterPicker());
    // 点击过滤器按钮触发过滤弹窗
    wrapper.find('button').at(0).simulate('click');

    // 此时这两个按钮应该为 disabled 状态
    const buttons = wrapper.find('button[disabled]');
    expect(buttons.at(0).text()).toBe('添加过滤条件');
    expect(buttons.at(1).text()).toBe('确 认');

    wrapper.find('.gio-picker-trigger').simulate('click');
    // 正确渲染 1 个属性
    expect(wrapper.find('.cascader-menu-item-inner')).toHaveLength(1);
  });

  it('cancel', () => {
    const wrapper = mount(getFilterPicker());
    wrapper.find('button').simulate('click');
    wrapper.find('.gio-dropdown').find('.filter-contidion-footer button').at(0).simulate('click');
    expect(wrapper.find('.gio-dropdown')).toHaveLength(0);
  });

  it('onConfirm', async () => {
    const onConfirm = jest.fn();
    let wrapper = null;
    wrapper = mount(getFilterPicker({ onConfirm }));
    wrapper.find('button').at(0).simulate('click');
    await sleep(100);
    wrapper.find('.gio-picker-trigger').simulate('click');
    expect(wrapper.find('.gio-property-picker-trigger').hasClass('open')).toBe(true);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('Dropdown').at(1).find('.cascader-menu-item-inner').at(0).simulate('click');
    await sleep(100);
    wrapper.update();
    expect(wrapper.find('.gio-picker-trigger').text()).toBe('test_int_name');
    expect(wrapper.find('.filter-condition_select').exists()).toBe(true);
    wrapper.find('.filter-condition_select').simulate('click');
    wrapper.find('.gio-select').simulate('click');
    wrapper.find('.gio-select-dropdown .gio-select-option').at(1).simulate('click');
    wrapper.find('Dropdown').at(2).find('.filter-contidion-footer').find('button').at(1).simulate('click');
    wrapper.find('Dropdown').at(0).find('.filter-contidion-footer').find('button').at(1).simulate('click');
    expect(onConfirm).toBeCalledTimes(1);
  });
});
