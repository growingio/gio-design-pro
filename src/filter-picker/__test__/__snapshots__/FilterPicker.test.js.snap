// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterPicker should match FilterPicker base snapshot. 1`] = `
<button
  class="gio-btn gio-btn-text gio-btn-small gio-btn-icon-only"
  style="cursor:pointer"
  type="button"
>
  <span
    aria-label="filter-outlined"
    class="gio-icon"
    role="img"
  >
    <svg
      fill="currentColor"
      height="16px"
      viewBox="0 0 14 14"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs />
      <g>
        <g>
          <path
            d="M6 6.46L5.56 6zM7.5 13a.51.51 0 00.5-.5V6.86a1 1 0 01.28-.69l4.16-4.32a.55.55 0 00.1-.55.51.51 0 00-.46-.3H1.92a.51.51 0 00-.46.3.53.53 0 00.1.55l4 4.15.16.17a1 1 0 01.28.69v5.64a.51.51 0 00.5.5zM9 13a1 1 0 01-1 1H6a1 1 0 01-1-1V7.28a1 1 0 00-.27-.69L.89 2.53A1.5 1.5 0 012 0h10a1.5 1.5 0 011.09 2.53L9.27 6.59a1 1 0 00-.27.69z"
            fill="currentColor"
            fill-rule="evenodd"
            id="filter-outlined_svg__filter"
          />
        </g>
      </g>
    </svg>
  </span>
</button>
`;

exports[`FilterPicker when use children, should render it correctly 1`] = `
<div
  class=""
  style="cursor:pointer"
>
  过滤选择器
</div>
`;
